pkg/ebpf/c/tracee.bpf.c

```diff
@@ -970,12 +983,152 @@ statfunc int walk_mod_tree(program_data_t *p, struct rb_node *root, int idx)
+    }
+
+    bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_PROC);
+
+    return 0;
+}
+
+SEC("uprobe/lkm_seeker_modtree_2_tail")
+int lkm_seeker_modtree_2_tail(struct pt_regs *ctx)
+{
+    program_data_t p = {};
+    if (!init_tailcall_program_data(&p, ctx))
+        return -1;
+
+    struct rb_node *curr = get_mod_tree_rb_node();
+    if (curr == NULL)
+        return -1;
+
+    int idx = get_mod_tree_rb_idx();
+    if (idx == -1)
+        return -1;
+
+    u32 flags = HISTORY_SCAN_FINISHED;
+    int ret = mod_tree_loop(&p, curr, idx);
+    if (ret == 0) {
+        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_SUCCESSFUL, &flags, &p);
+    } else if (ret == HID_MOD_UNCOMPLETED_ITERATIONS) { // Continue doing more iterations
+        bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_MODTREE_3);
+    } else {
+        tracee_log(ctx, BPF_LOG_LVL_WARN, BPF_LOG_ID_HID_KER_MOD, ret);
+        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_FAILURE, &flags, &p);
+    }
+
+    bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_PROC);
+
+    return -1;
+}
+
+SEC("uprobe/lkm_seeker_modtree_3_tail")
+int lkm_seeker_modtree_3_tail(struct pt_regs *ctx)
+{
+    program_data_t p = {};
+    if (!init_tailcall_program_data(&p, ctx))
+        return -1;
+
+    struct rb_node *curr = get_mod_tree_rb_node();
+    if (curr == NULL)
+        return -1;
+
+    int idx = get_mod_tree_rb_idx();
+    if (idx == -1)
+        return -1;
+
+    u32 flags = HISTORY_SCAN_FINISHED;
+    int ret = mod_tree_loop(&p, curr, idx);
+    if (ret == 0) {
+        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_SUCCESSFUL, &flags, &p);
+    } else if (ret == HID_MOD_UNCOMPLETED_ITERATIONS) { // Continue doing more iterations
+        bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_MODTREE_4);
+    } else {
+        tracee_log(ctx, BPF_LOG_LVL_WARN, BPF_LOG_ID_HID_KER_MOD, ret);
+        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_FAILURE, &flags, &p);
+    }
+
+    bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_PROC);
+
+    return -1;
+}
+
+SEC("uprobe/lkm_seeker_modtree_4_tail")
+int lkm_seeker_modtree_4_tail(struct pt_regs *ctx)
+{
+    program_data_t p = {};
+    if (!init_tailcall_program_data(&p, ctx))
+        return -1;
+
+    struct rb_node *curr = get_mod_tree_rb_node();
+    if (curr == NULL)
+        return -1;
+
+    int idx = get_mod_tree_rb_idx();
+    if (idx == -1)
+        return -1;
+
+    u32 flags = HISTORY_SCAN_FINISHED;
+    int ret = mod_tree_loop(&p, curr, idx);
+    if (ret == 0) {
+        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_SUCCESSFUL, &flags, &p);
+    } else {
+        tracee_log(ctx, BPF_LOG_LVL_WARN, BPF_LOG_ID_HID_KER_MOD, ret);
+        lkm_seeker_send_to_userspace((struct module *) HISTORY_SCAN_FAILURE, &flags, &p);
+    }
+
+    bpf_tail_call(ctx, &prog_array, TAIL_HIDDEN_KERNEL_MODULE_PROC);
+
+    return -1;
+}
```

Comment:
Why do we need 4 different programs that do the same thing?
Can't you use the context map (that you added) and use one program that tail calls itself recursively?

Reply:
Great point.. Changed that.